src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+    encode_info info(encode, true);
+    info.add_input(op1_val);
+    return info;
+}
+
+// Encode common fields in convolution:
+//
+// |----- 16 bits -----|----- 16 bits -------|----- 8 bits -----|----- 8 bits-----|----- 16 bits
+// -----| |     opcode        | 1st operand hash id |   filter size    |  kernel size    |
+// 0x0000        |
+
+encode_info encode_conv_common(instruction_ref ins, ins2_val& instr2_value, unsigned opcode)
+{
+    encode_info info = encode_common(ins, instr2_value, opcode);
+    if(!info.is_valid())
+        return info;
+    key_type encode     = info.get_key();
+    instruction_ref op2 = ins->inputs().at(1);
+    auto lens           = op2->get_shape().lens();
+    auto filter         = lens[lens.size() - 2];
+    auto kernel         = lens[lens.size() - 1];
+    if((filter < (static_cast<unsigned>(1) << filter_bits)) &&
+       (kernel < (static_cast<unsigned>(1) << kernel_bits)))
+    {
+        encode |= (filter << filter_shift_count());
+        encode |= (kernel << kernel_shift_count());
+        info.set_key(encode);
+        return info;
+    }
+    else
+    {
+        return encode_info(0, false);
+    }
+}
+
+bool horizontal_fusion_impl::is_conv(instruction_ref ins)
+{
+    if(op_flag.find(ins->name()) == op_flag.end())
+        return false;
+    return (op_flag[ins->name()] == 1);
+}
+
+bool horizontal_fusion_impl::is_concat(instruction_ref ins) { return (ins->name() == "concat"); }
+
+// Hash given instruction.
+hash_value_ptr horizontal_fusion_impl::hash(instruction_ref ins)
+{
+    if(op_registry.find(ins->name()) == op_registry.end())
+        return nullptr;
+
+    encoder encode_func    = op_registry.at(ins->name());
+    unsigned opcode        = hash_opcode(ins);
+    encode_info encode_val = encode_func(ins, instr2_value, opcode);
+    if(!encode_val.is_valid())
+        return nullptr;
+
+    key_type key            = encode_val.get_key();
+    hash_value_ptr hash_val = nullptr;
+
+    if(encode2_value.find(key) != encode2_value.end())
+    {
+        hash_val = encode2_value[key];
+        add_instr(hash_val->id);
+        instr2_value[ins] = hash_val;
+    }
+    else
+    {
+        hash_val           = &(create_value(ins));
+        encode2_value[key] = hash_val;
+    }
+    for(auto&& input : encode_val.get_inputs())
+    {
+        add_input(hash_val->id, input);
+        add_output(input->id, hash_val);
+    }
+    return hash_val;
+}
+
+hash_value& horizontal_fusion_impl::create_value(instruction_ref ins)
+{
+    unsigned id = static_cast<unsigned>(values.size());
+    values.push_back(hash_value{id, cur_point});
+    hash_value& val = get_value(id);
+    add_instr(id);
+    instr2_value[ins] = &val;
+    return val;
+}
+
+void horizontal_fusion_impl::process(instruction_ref ins)
+{
+    // Do not hash literals.
+    if(ins->name() == "@literal")
+        return;
+    if(instr2_hash.find(ins) != instr2_hash.end())
```

Comment:
What does the boolean in `instr2_hash` mean? It seems to be only assigned to `true`, which this could be written as a set.

Reply:
instr2_hash hashes an instruction to a boolean to indicate where an instruction needs to be hashed.

Reply:
> instr2_hash hashes an instruction to a boolean to indicate where an instruction needs to be hashed.

Couldn't this just be a set instead?

And if its boolean why do you use find? You could just write `instr2_hash[ins]`?

Reply:
It means the instruction needs to be hashed.